proposals/module-types/Example-SharedEverythingDynamicLinking.md

```diff
@@ -0,0 +1,524 @@
+# Shared-Everything Dynamic Linking Example
+
```

Comment:
This is very cool, and I would be happy if tools like wasm-ld could one day emit modules in this format when building and using dynamic libraries.

One concern is that it seems like you have only addressed function symbols.   Modules in llvm also have global data and corresponding data symbols.    In the current llvm+emscripten model we deal with this in the following way: Each shared library has its own data and elem segments which are created at static link time.  The key is that these segments have dynamic base addresses based on wasm global which are imported as  `__table_base` and `__memory_base`. 

Data symbols are then imported and exported just like function symbols in this proposal.  When imported, data address are expected to be absolute.  When exported data addresses are assumed to be relative to the module's `__memory_base`.

In addition to this basic use of data symbols there is also the problem the relocations which are required in the data section.   Unlike with the code section we have not found a way to avoid these.   For example:

```
extern int foo;
int* bar = &foo;
````

The result of compiling this code into a shared library is that it allocates 4 bytes of static data along with an associated relocation entry.     In the current llvm+emscripten model these relocations are turned into generated code that runs during the start function so they are effectively applied by the module itself, rather than some outside dynamic linker.   This simplifies the dynamic linker at the expense of some codegen performed by wasm-ld.  Its also means we don't have to spec any kind of format for relocations in the shared library format, since we leave it all up to the module itself to self-relocate on startup.

Reply:
You're right, this example doesn't mention data symbols, but also, yes, they could be handled quite symmetrically to functions (particularly exported function pointer identities).  Do you think it's worth adding another segment below the "Function Pointer Identity" section mentioning these cases and saying it's symmetric?  Are there any hard cases you think aren't addressed by such a scheme?

Reply:
I certainly think that it worth specifying how data symbols might imported and exported in this scheme.   This I think this will naturally force us to define how modules can include and use their own static data.

Perhaps we could illustrate this by adding a string constant to the example that one of the libraries exports to the main program as a data symbol?

I fear it might add a fair amount of complexity, and I don't want to block this PR if you feel like you want to get something landed and then iterate?     I'm saying that partly because I am aware my comments are coming quite a late in the discussion here.

Reply:
Yeah, happy to add another little section with an example.  I'll try to get to that tomorrow.